torch/csrc/distributed/rpc/utils.cpp

```diff
@@ -20,6 +21,27 @@ namespace torch {
 namespace distributed {
 namespace rpc {
 
+RPCErrorType getRPCErrorType(const FutureMessage& fm) {
+  TORCH_INTERNAL_ASSERT(
+      fm.hasError(),
+      "FutureMessage passed to getRPCErrorType does not have an error.");
+  auto timeoutErrors =
+      RpcAgent::getCurrentRpcAgent()->getTimeoutErrorDescription();
+  timeoutErrors.emplace_back("failed intentionally");
+
+  auto err = std::string(fm.error()->what());
```

Comment:
I like that we're trying to distinguish between failure causes here.
Though scanning strings for a vector of possible error cases feels like it might be a bit cumbersome, and unnecessarily widen the RpcAgent api.

As a potential [half-baked] alternate suggestion, if we need the scan strings from the future error, maybe we can establish a convention where we tag the error string?

e.g. maybe something like "RpcErr:code"? 
    "RpcErr:1 RPC ran for more than {} milliseconds and timed out"
    "Host {} is down (RpcErr:2)"
    "Host {} is overloaded (RpcErr:3)"  ?
    etc.

(The tag needn't be at the beginning,  we can just strstr for the first occuring place, and if there's a parsable int, use that. Otherwise, categorize as unknown).

   "Rpc

Reply:
Sure, this sounds reasonable. I think we can take the error type from the enum when we are handling the error (since we know what the error is) and then format that into the error message string. Then in this function, instead of parsing for substrings, we could instead parse out the error code as you suggested.

Reply:
@jjlilley I updated the diff so that common errors (when we can detect them in the RPC agents) are tagged with the function `makeRPCError`, and this parsing logic is updated to attempt to get the tagged error from the error message.